### DISCUSSION TITLE
SignalDB as web assembly for improved performance.

### DISCUSSION THREAD

**Discussion by michaelkrog** [maxnowack/signaldb#828] [Ideas]
I know that it is a major idea to raise and that it would require pretty much a rewrite of the core of signaldb..

The reason I raise the idea anyway, is that https://github.com/maxnowack/signaldb/issues/397 suggest the ability to move things of the main thread for improved performance. Having signaldb as web assembly might improve it even further. Using Rust for the purpose might even push the boundaries for stable performance even more, as it has no garbage collection to run now and then (Check out [Rust for Wasm](https://www.rust-lang.org/what/wasm)).

Also see this result from a recent paper on efficiency of different programming languages:
![Skærmbillede 2024-08-06 kl  11 54 04](https://github.com/user-attachments/assets/5b6d3ee3-2ba3-4e87-8531-143e74885a24)

https://haslab.github.io/SAFER/scp21.pdf

**Comment by maxnowack** [comment]
I really like the idea and had also already thought about it myself. I have already tried Rust in another one of my projects (github.com/maxnowack/blend), but then decided to first develop an initial version in TypeScript to better understand if the basic idea would work at all. Refactoring to Rust is still pending here as well. The scope of the project is a bit smaller than that of SignalDB, so it will be more suitable for learning Rust. Once I am more comfortable with Rust, I plan to rewrite SignalDB as well. However, I think this will take some more time.
